### PR TITLE
Add TabTransformer code example

### PR DESCRIPTION
**Pull request by ksalama**
Hello @fchollet - Kindly review this PR. 

This example demonstrates how to do structured data classification using [TabTransformer](https://arxiv.org/abs/2012.06678), a deep learning model for structured data data. The TabTransformer is built upon self-attention based Transformers. The Transformer layers transform the embeddings of categorical features into robust contextual embeddings to achieve higher predictive accuracy.

### CODE REVIEW

File: examples/structured_data/tabtransformer.py [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+        csv_file_path,
+        batch_size=batch_size,
+        column_names=CSV_HEADER,
+        column_defaults=COLUMN_DEFAULTS,
+        label_name=TARGET_FEATURE_NAME,
+        num_epochs=1,
+        header=False,
+        na_value="?",
+        shuffle=shuffle,
+    ).map(prepare_example)
```
Comment (fchollet) [inline examples/structured_data/tabtransformer.py:198]: Make sure to use `num_parallel_calls`

File: examples/structured_data/tabtransformer.py [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+"""
+
+
+def encode_inputs(inputs, embedding_dims):
+
+    encoded_categorical_feature_list = []
+    numerical_feature_list = []
+
+    for feature_name in inputs:
+        if feature_name in CATEGORICAL_FEATURE_NAMES:
```
Comment (fchollet) [inline examples/structured_data/tabtransformer.py:273]: Likewise, this function will be more reusable if it only relies on function arguments rather than global constants

File: examples/structured_data/tabtransformer.py [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+
+"""
+## Implement data reading pipeline
+
+We define an input function that reads and parses the file, then converts features
+and labels into a[`tf.data.Dataset`](https://www.tensorflow.org/guide/datasets)
+for training or evaluation.
+"""
+
+from tensorflow.keras.layers import StringLookup
```
Comment (fchollet) [inline examples/structured_data/tabtransformer.py:174]: You can just access it via `layers`, no need to import the symbol itself

File: examples/structured_data/tabtransformer.py [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+    # encode features.
+    encoded_categorical_feature_list, numerical_feature_list = encode_inputs(
+        inputs, embedding_dims
+    )
+    # Concatenate all features.
+    features = layers.concatenate(
+        encoded_categorical_feature_list + numerical_feature_list
+    )
+    # Compute Feedforward layer units.
+    ff_units = [features.shape[-1]]
```
Comment (fchollet) [inline examples/structured_data/tabtransformer.py:347]: Use fully spelled-out variable names

File: examples/structured_data/tabtransformer.py [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+
+    # Create several feedforwad layers with skip connections.
+    for layer in range(num_mlp_blocks):
```
Comment (fchollet) [inline examples/structured_data/tabtransformer.py:350]: `layer_index`?

File: examples/structured_data/tabtransformer.py [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+        column_embedding = layers.Embedding(
+            input_dim=num_columns, output_dim=embedding_dims
+        )
+        column_indices = tf.range(start=0, limit=num_columns, delta=1)
+        encoded_categorical_features = encoded_categorical_features + column_embedding(
+            column_indices
+        )
+
+    # Create multiple layers of the Transformer block.
+    for block in range(num_transformer_blocks):
```
Comment (fchollet) [inline examples/structured_data/tabtransformer.py:449]: Likewise it would be useful to have an indication that `block` is a counter (e.g. `block_index` or something) rather than an actual block/layer object

**Review by fchollet — CHANGES_REQUESTED** [review]
Fantastic example demonstrating a very useful architecture! Thanks for the PR.

File: examples/structured_data/tabtransformer.py (hunk 5 above)
Comment (ksalama) [reply examples/structured_data/tabtransformer.py:350]: Done

File: examples/structured_data/tabtransformer.py (hunk 6 above)
Comment (ksalama) [reply examples/structured_data/tabtransformer.py:449]: Done

File: examples/structured_data/tabtransformer.py (hunk 4 above)
Comment (ksalama) [reply examples/structured_data/tabtransformer.py:347]: Done

File: examples/structured_data/tabtransformer.py (hunk 3 above)
Comment (ksalama) [reply examples/structured_data/tabtransformer.py:174]: Done

File: examples/structured_data/tabtransformer.py (hunk 1 above)
Comment (ksalama) [reply examples/structured_data/tabtransformer.py:198]: Done. The parameter is called `num_parallel_reads`.

File: examples/structured_data/tabtransformer.py [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+# A list of the labels of the target features.
+TARGET_LABELS = [" <=50K", " >50K"]
+
+"""
+## Configure the hyperparameters
+
+The hyperparameters includes model architecture and training configurations.
+"""
+
+learning_rate = 0.001
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/tabtransformer.py:151]: Done

File: examples/structured_data/tabtransformer.py [hunk 8]
```diff
@@ -0,0 +1,537 @@
...
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2022/01/31
+Last modified: 2022/01/31
+Description: Using contextual embeddings for structured data classification.
+"""
+
+"""
+## Introduction
+
+This example demonstrates how to do structured data classification using [TabTransformer](https://arxiv.org/abs/2012.06678),
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/structured_data/tabtransformer.py]: Done

File: examples/structured_data/tabtransformer.py (hunk 2 above)
Comment (ksalama) [reply examples/structured_data/tabtransformer.py:282]: I fixed this in the hyperparameters. However, I would keep the dataset metadata as global constants.
LMK if you are OK with this.

**Comment by ksalama** [comment]
@fchollet Thank you so much for your comments. I addressed them and updated the PR.

File: examples/structured_data/tabtransformer.py (hunk 1 above)
Comment (fchollet) [reply examples/structured_data/tabtransformer.py:199]: Ok, I actually meant the `num_parallel_calls` argument in `map()` -- otherwise you will call the function serially, which is slower

Comment (ksalama) [reply examples/structured_data/tabtransformer.py]: Ah OK. Added it now

**Comment by ksalama** [comment]
@fchollet - Please LMK if you have further comments. Cheers!

**Review by fchollet — APPROVED** [review]
Awesome, thank you for the update! Please add the generated files.

**Comment by ksalama** [comment]
@fchollet Thank you so much for approving the PR. I have added the generated files.

Another interesting one is coming after this one is merged ;-)

**Comment by fchollet** [comment]
> Another interesting one is coming after this one is merged

Awesome, looking forward!
